auth/service.cc

```diff
@@ -627,4 +633,66 @@ future<std::vector<permission_details>> list_filtered_permissions(
     });
 }
 
+future<> migrate_to_auth_v2(cql3::query_processor& qp, ::service::raft_group0_client& g0, start_operation_func_t start_operation_func, abort_source& as) {
+    // FIXME: if this function fails it may leave partial data in the new tables
+    // that should be cleared
+    auto gen = [&qp] (api::timestamp_type& ts) -> mutations_generator {
+        for (const auto& cf_name : std::vector<sstring>{
+            "roles", "role_members", "role_attributes", "role_permissions"}) {
+            schema_ptr schema;
+            try {
+                schema = qp.db().find_schema(meta::legacy::AUTH_KS, cf_name);
+            } catch (data_dictionary::no_such_column_family&) {
+                continue; // some tables might not have been created if they were not used
```

Comment:
Could you provide examples of such tables? Can they only be created due to user action, or can they be created automatically?

Reply:
Continuing the line of thought, can data in those tables be modified on a whim by Scylla? I recall something about superuser role being created after a short delay on startup.

If it is possible, then it will be very important to document those conditions in the upgrade procedure guide and provide instructions on how to verify them.

Hopefully, there are no "random" modifications like that, i.e. that users can't predict reliably when they happen - if they are, then the procedure should be made more resilient to them.

Reply:
role_permissions is such table, and in general those tables are not exactly system, they are created dynamically on startup, perhaps on some older versions it was even different. 

they are no random modifications, each is effect of user's statement except on first startup. but I think we don't need to handle case where user is concurrently performing migration and configuring auth the first time especially if we plan to explicitly say auth modifications and schema changes are not allowed during migration.

> provide instructions on how to verify them

since those are admin commands I assume whoever is doing the upgrade is in full control, I guess we can implement some metric checks or something but that's another discussion if it's needed

Reply:
The one who is in full control needs to know what to do. And we, the developers, need to tell them what the safe conditions are. We can update documentation as a follow-up (the upgrade page is still to be written by @piodul), but we need to understand the conditions right now, so that we'll know what to write.

Reply:
> role_permissions is such table, and in general those tables are not exactly system, they are created dynamically on startup, perhaps on some older versions it was even different.

If it's dynamically created on startup, why not just assume that the table is created?

> they are no random modifications, each is effect of user's statement except on first startup. but I think we don't need to handle case where user is concurrently performing migration and configuring auth the first time especially if we plan to explicitly say auth modifications and schema changes are not allowed during migration.

I was only concerned about auth modifications which are out of control of the administrator.

As for the role_permissions table created at startup, considering that upgrade is being done only after a rolling upgrade of an existing legacy cluster, I suppose we can assume that the table was created long time ago.

> since those are admin commands I assume whoever is doing the upgrade is in full control, I guess we can implement some metric checks or something but that's another discussion if it's needed

I don't think there is any need for metrics etc., it's sufficient that we tell administrators not to do dangerous things during upgrade. Later, when we make the upgrade procedure fully safe to run automatically, the cluster itself will detect when it's safe to run and will prevent unsafe things from happening during upgrade.

Reply:
> If it's dynamically created on startup, why not just assume that the table is created?

it's dynamically created when certain auth is on, otherwise not created.

> we need to understand the conditions right now, so that we'll know what to write

Safe conditions are as I wrote above:
- no auth modifications
- no auth config changes
- no schema changes (due to auto-grant, auto-revoke)

Reply:
The new auth-v2 tables are all created non-conditionally, right? (Which is good)

Reply:
the condition is consistent topology feature flag, so they will be "created" (I just register them as system tables at startup) before migration starts i.e. also when we're still on auth-v1

Reply:
> the condition is consistent topology feature flag

do you mean the cluster feature, or the experimental-feature?
(The experimental-feature will be removed before release. The cluster feature will stay)

Reply:
experimental-feature

Reply:
Ok, so they will be created unconditionally